opendatakit/briefcase
# Briefcase analytics

This PR comes with Google Analytics tracking. The list of things we're tracking is:
- Open/close Briefcase
- Enter/leave all four tabs
- Success/failure at exporting forms
- Success/failure at pulling forms
- Success/failure at pushing forms
- Grant/revoke permission to track activity

Every hit is submitted with the following information attached:
- Briefcase version
- Unique User ID
- Anonymized IP address
- Operating System
- Screen resolution
- Briefcase window resolution

**Testing guidelines**
I guess one way to test this is to create a GA account and configure the GA Tracking ID in `build.gradle`. Then, browse the app with and without having checked the tracking consent on the Settings panel. You should see a lot of activity on the real-time report whenever tracking is enabled.

**About the User ID**
Briefcase will generate a User ID (a UUID) and store it using Java user preferences. If these preferences get removed, a new (different) User ID will be created.

**About the Operating System**
Google Analytics doesn't let us specify an OS manually. Instead, we are using three default user agent strings (one for each OS we want to track). This way, Google Analytics (hopefully) will be able to add OS info to each hit.

**About the GA Tracking ID**
I have created a GA account using my google account to be able to work this feature. For obvious reasons, this GA Tracking ID must not be used in the release versions of Briefcase.

#### What has been done to verify that this works as intended?
Manual tests on Linux, Mac, and Win10

#### Why is this the best possible solution? Were any other approaches considered?
We could have used the previous code but the GA library [was updated recently and had breaking changes](https://github.com/brsanthu/google-analytics-java).

#### Are there any risks to merging this code? If so, what are they?
Tracking uses network resources. Although all hits are sent asynchronously (meaning GA shouldn't affect negatively to the user experience), maybe on scenarios with low bandwidth or data support, users will be forced to remove their consents.
